packages/SwingSet/src/kernel/kernel.js

```diff
@@ -553,42 +630,57 @@ export default function buildKernel(kernelEndowments) {
       kernelKeeper.createStartingKernelState();
     }
 
-    function assignIdAndBuildVatManager(name, setup, options) {
-      const vatID = kernelKeeper.provideVatIDForName(name);
-      console.log(`Assigned VatID ${vatID} for Genesis vat ${name}`);
-      buildVatManager(vatID, name, setup, options);
+    // TESTONLY: we condition on vatAdminVatSetup and vatAdminDevSetup because
+    // some kernel tests don't care about vats and devices and so don't
+    // initialize properly. We should fix those tests.
+    if (vatAdminVatSetup) {
+      genesisVats.set('vatAdmin', { setup: vatAdminVatSetup, options: {} });
     }
 
-    // instantiate all other vats
+    // instantiate all vats
     for (const name of genesisVats.keys()) {
       const { setup, options } = genesisVats.get(name);
-      assignIdAndBuildVatManager(name, setup, options);
+      const vatID = kernelKeeper.provideVatIDForName(name);
+      console.log(`Assigned VatID ${vatID} for genesis vat ${name}`);
+      const manager = buildVatManager(vatID, name, setup);
+      ephemeral.vats.set(
+        vatID,
+        harden({
+          manager,
+          enablePipelining: Boolean(options.enablePipelining),
+        }),
+      );
     }
 
-    function setupDeviceManger(name, setup, endowments) {
-      const deviceID = kernelKeeper.provideDeviceIDForName(name);
-      const manager = buildDeviceManager(deviceID, name, setup, endowments);
-      // the vat record is not hardened: it holds mutable next-ID values
-      ephemeral.devices.set(deviceID, {
-        manager,
-      });
+    function getVatControlFns() {
+      return harden({ create: createVatDynamically /* vatStats, terminate */ });
+    }
+
+    if (vatAdminDevSetup) {
+      const params = {
+        setup: vatAdminDevSetup,
+        endowments: { getVatControlFns },
```

Comment:
this could be simplified by providing `create` (+etc) as an endowment, rather than a function which returns `create`

Reply:
Thanks.